### PR TITLE
Add {ReadOnly}Memory.Span tests when memory contains invalid values

### PR DESCRIPTION
**Pull request by stephentoub**
Related to https://github.com/dotnet/coreclr/pull/17452#discussion_r179748571
cc: @ahsonkhan, @jkotas

### CODE REVIEW

File: src/System.Memory/tests/TestHelpers.cs [hunk 1]
```diff
@@ -398,6 +399,22 @@ public static IEnumerable<object[]> StringSlice3ArgTestOutOfRangeData
...
+            object boxedMemory = mem;
+
+            typeof(Memory<T>).GetField("_object", BindingFlags.Instance | BindingFlags.NonPublic).SetValue(boxedMemory, obj);
+            typeof(Memory<T>).GetField("_index", BindingFlags.Instance | BindingFlags.NonPublic).SetValue(boxedMemory, offset);
+            typeof(Memory<T>).GetField("_length", BindingFlags.Instance | BindingFlags.NonPublic).SetValue(boxedMemory, length);
+
+            return (Memory<T>)boxedMemory;
+        }
+
+        /// <summary>Creates a <see cref="Memory{T}"/> with the specified values in its backing field.</summary>
```
Comment (ahsonkhan) [inline src/System.Memory/tests/TestHelpers.cs]: nit: Creates a ReadOnlyMemory{T}

File: src/System.Memory/tests/Memory/Span.cs [hunk 2]
```diff
@@ -106,5 +106,28 @@ public static void SpanFromDefaultMemory()
             Assert.True(spanObject.SequenceEqual(default));
         }
 
+        [Fact]
+        public static void TornMemory_Array_SpanThrowsIfOutOfBounds()
```
Comment (ahsonkhan) [inline src/System.Memory/tests/Memory/Span.cs]: Should we add similar tests for Pin as well?

Comment (stephentoub) [reply src/System.Memory/tests/Memory/Span.cs]: Yes, but I don't think such tests would currently pass. We need to fix the code as well.

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Pull request merged**